tempesta_fw/http_match.c

```diff
@@ -201,27 +202,99 @@ match_hdr(const TfwHttpReq *req, const TfwHttpMatchRule *rule)
 	return hdr_val_eq(req, id, rule->arg.str, rule->arg.len, flags);
 }
 
+#define _MOVE_TO_COND(p, end, cond)			\
+	while((p) < (end) && !(cond)) {			\
+		(p)++;					\
+	}
+
+/* It would be hard to apply some header-specific rules here, so ignore
+ * case for all headers according to the robustness principle.
+ */
 static bool
 match_hdr_raw(const TfwHttpReq *req, const TfwHttpMatchRule *rule)
 {
-	TfwStr *hdr;
 	int i;
 	tfw_str_eq_flags_t flags = map_op_to_str_eq_flags(rule->op);
 
-	/* It would be hard to apply some header-specific rules here, so ignore
-	 * case for all headers according to the robustness principle. */
-	flags |= TFW_STR_EQ_CASEI;
-
 	for (i = 0; i < req->h_tbl->size; ++i) {
+		const TfwStr *hdr, *dup, *end, *chunk;
+		const char *c, *cend, *p, *pend;
+		char prev;
+		short cnum;
+
 		hdr = &req->h_tbl->tbl[i];
-		if (TFW_STR_EMPTY(hdr))
+		if (TFW_STR_EMPTY(hdr)) {
 			continue;
+		}
+
+		TFW_STR_FOR_EACH_DUP(dup, hdr, end) {
+			/* Initialize  the state - get the first chunk. */
+			p = rule->arg.str;
+			pend = rule->arg.str + rule->arg.len;
+			cnum = 0;
+			chunk = TFW_STR_CHUNK(dup, 0);
+			if (!chunk) {
+				return p == NULL;
+			}
+			c = chunk->ptr;
+			cend = chunk->ptr + chunk->len;
+
+#define _TRY_NEXT_CHUNK(ok_code, err_code)		\
+	if (unlikely(c == cend))	{		\
+		++cnum;					\
+		chunk = TFW_STR_CHUNK(dup, cnum); 	\
+		if (chunk) {				\
+			c = chunk->ptr;			\
+			cend = chunk->ptr + chunk->len; \
+			ok_code;			\
+		} else {				\
+			err_code;			\
+		}					\
+	}
 
-		/* TODO: handle LWS* between header and value for raw headers.
-		 * (currently "X-Hdr:foo" is not equal to "X-Hdr: foo").
-		 */
-		if (tfw_str_eq_cstr(hdr, rule->arg.str, rule->arg.len, flags))
-			return true;
+state_common:
+			while (p != pend && c != cend) {
+				if (tolower(*p) != tolower(*c)) {
+					/* If the position of the strings have a
+					 * different number of space characters,
+					 * or one of the strings have space
+					 * characters after ':', missing their
+					 */
+					if (unlikely((
+						(isspace(*p) &&
+							(prev == *p ||
+								prev == ':'))
+						||
+						(isspace(*c) &&
+							(prev == *c ||
+								prev == ':'))
```

Comment:
```prev``` is checked against ':' twice. You can easily rewrite the expression with only one check. Meantime there is nothing bad to do couple of simple checks with the same body instead of this one large condition - compiler also generates just set of conditional jumps.